what4/src/What4/Concretize.hs

```diff
@@ -0,0 +1,108 @@
+-----------------------------------------------------------------------
+-- |
+-- Module           : What4.Concretize
+-- Description      : Concretize values
+-- Copyright        : (c) Galois, Inc 2024
+-- License          : BSD3
+-- Maintainer       : Langston Barrett <langston@galois.com>
+-- Stability        : provisional
+-----------------------------------------------------------------------
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeOperators #-}
+
+module What4.Concretize
+  ( GroundingFailure(..)
+  , groundFromModel
+  , ConcretizationFailure(..)
+  , concretize
+  ) where
+
+import qualified What4.Expr.Builder as WEB
+import qualified What4.Expr.GroundEval as WEG
+import qualified What4.Interface as WI
+import qualified What4.Protocol.Online as WPO
+import qualified What4.Protocol.SMTWriter as WPS
+import qualified What4.SatResult as WSat
+
+-- | Reasons why attempting to resolve a symbolic expression as ground can fail.
+data GroundingFailure
+  = SolverUnknown
+    -- ^ Querying the SMT solver yielded @UNKNOWN@.
+  | UnsatInitialAssumptions
+    -- ^ Querying the SMT solver for an initial model of the expression failed
+    -- due to the initial assumptions in scope being unsatisfiable.
+  deriving Show
+
+-- | Get a 'WEG.GroundValue' for a 'WI.SymExpr' by asking the solver for a
+-- model.
```

Comment:
Since `groundFromModel` and `concretize` have nearly the same type signature, it would be worth contrasting each function to make it clearer when you'd pick one over the other.

Reply:
Yep, good point. Done!